crossdock/server/server.py

```diff
@@ -77,25 +83,41 @@ def head(self):
 def make_app(server):
     return tornado.web.Application(
         [
-            (r"/", MainHandler),
-            (r"/start_trace", MainHandler, (dict(server=server, method=Server.start_trace))),
-            (r"/join_trace", MainHandler, (dict(server=server, method=Server.join_trace))),
+            (r'/', MainHandler),
+            (r'/start_trace', MainHandler, (dict(server=server, method=Server.start_trace))),
+            (r'/join_trace', MainHandler, (dict(server=server, method=Server.join_trace))),
         ], debug=True)
 
 
-# TChannel handlers
-def tchannelNotImplResponse():
-    return TraceResponse(notImplementedError="python <=> tchannel not implemented")
+def make_tchannel(port):
+    global tchannel
+    tchannel = TChannel('python', hostport='localhost:%d' % port, trace=True)
+
+    service = get_thrift_service(service_name='python')
+
+    @tchannel.thrift.register(service.TracedService, method='joinTrace')
+    @tornado.gen.coroutine
+    def join_trace(request):
+        join_trace_request = request.body.request or None
+        response = yield handle_downstream_request(join_trace_request.downstream)
+        raise tornado.gen.Return(response)
+
+    @tornado.gen.coroutine
+    def handle_downstream_request(downstream):
+        span = get_current_span()
+        observed_span = observed_span_to_thriftrw(service, get_observed_span(span))
 
+        trace_response = TraceResponse(span=observed_span, notImplementedError='')
 
-@tchannel.thrift.register(service.TracedService)
-def startTrace(request):
-    return trace_response_to_thriftrw(service, tchannelNotImplResponse())
+        if downstream:
+            downstream_trace_resp = yield call_downstream(span, downstream)
+            observed_span = observed_span_to_thriftrw(service, downstream_trace_resp.span)
+            downstream_trace_resp = TraceResponse(span=observed_span, notImplementedError='')
+            trace_response.downstream = trace_response_to_thriftrw(service, downstream_trace_resp)
```

Comment:
this method is only called from tchannel join_trace handler, so I think there's only one form the request would come in. My preference is to have only a single method `prepare_response`, instead of two as you have, and have that method always deal with the same type in and out, most likely json classes. Then if that method is called from JSON handlers, you don't need to convert anything, otherwise the handler first converts thrift->json, calls prepare_response(), then converts result to thrift. Clean separation, DRY, no confusion which type is in or out.

You can see this in tchannel-python/crossdock server.